Haskell/Haskell.sublime-syntax

```diff
@@ -1,407 +1,1388 @@
+        5: constant.character.escape.control.haskell
+
+###[ KEYWORDS AND OPERATORS ]##################################################
+
+  keywords:
+    - match: (?:mdo|do|in|rec|where){{break}}
       scope: keyword.control.context.haskell
-    - match: \b(?:newtype|type)\b
-      scope: keyword.declaration.type.haskell
-    - match: \b(?:data)\b
-      scope: keyword.declaration.data.haskell
-    - match: \b(?:deriving)\b
-      scope: keyword.declaration.data.haskell
-    - match: \b(?:case|of)\b
+    - match: (?:case|of){{break}}
       scope: keyword.control.conditional.select.haskell  # the construct is commonly called "select"
-    - match: \b(?:let|where)\b
+    - match: if{{break}}
+      scope: keyword.control.conditional.if.haskell
+    - match: then{{break}}
+      scope: keyword.control.conditional.then.haskell
+    - match: else{{break}}
+      scope: keyword.control.conditional.else.haskell
+    - match: let{{break}}
       scope: keyword.declaration.variable.haskell
-    - match: \breturn\b
-      scope: keyword.control.flow.return.haskell
-    - match: \b(?:default|otherwise)\b
-      scope: keyword.other.haskell
-    - match: \b(?:(if)|(then)|(else))\b
-      captures:
-        1: keyword.control.conditional.if.haskell
-        2: keyword.control.conditional.then.haskell
-        3: keyword.control.conditional.else.haskell
-
-  operator:
-    - match: (`)[a-zA-Z_'.\d]+(`)
-      # Haskell allows any ordinary function application (elem 4 [1..10])
-      # to be rewritten as an infix expression (4 `elem` [1..10])."
-      scope: keyword.operator.function.infix.haskell
-      captures:
-        1: punctuation.definition.function.begin.haskell
-        2: punctuation.definition.function.end.haskell
-    - match: '\binfix[lr]?\b'
+    - match: proc{{break}}
+      scope: keyword.declaration.function.haskell
+    # 4.4.2 Fixity Declarations
+    - match: infix[lr]?{{break}}
+      scope: keyword.declaration.fixity.haskell
+
+  operators:
+    - include: sequence-separators
+    - include: big-arrow-operators
+    - include: left-arrow-operators
+    - include: right-arrow-operators
+    - include: prefix-parens-operators
+    - include: infix-quoted-operators
+    # Match all not otherwise matched single quotes as promoition operator
+    # Note: Found in real world code but not in specs so far.
+    - match: \'
       scope: keyword.operator.haskell
-    - match: \b(\d+)(?:(\.)(\d+)([eE][-+]?\d+)?|([eE][-+]?\d+))\b
-      scope: meta.number.float.decimal.haskell
-      captures:
-        1: constant.numeric.value.haskell
-        2: punctuation.separator.decimal.haskell
-        3: constant.numeric.value.haskell
-        4: constant.numeric.value.exponent.haskell
-        5: constant.numeric.value.exponent.haskell
-    - match: '{{operator_infix}}'
+    - match: '{{operator_double_colon}}'
+      scope: keyword.operator.double-colon.haskell
+    - match: '{{operator_symbol}}'
```

Comment:
I'd appreciate a separate match for the `=` assignment operator.

Reply:
The old syntax does scope `=` as `keyword.operator.haskell`, too. By not applying semantic scoping in general I am not sure about making such exceptions.

Reply:
Maybe @mitranim can confirm whether `=` cannot be used anywhere else outside of an assignment?

Reply:
To the best of my knowledge (which is probably incomplete), Haskell uses `=` for exactly one purpose: declarations. Haskell doesn't have an "assignment" as understood in other languages, because it doesn't allow to _reassign_ names. In root scope:

```hs
funcName <params_patterns_guards> = <expressions>
data TypeName = <variants>
```

In local scope:

```hs
funcName = do
  let localFunc <params_patterns_guards> = <expressions>
  localFunc

funcName = localFunc
  where
    localFunc <params_patterns_guards> = <expressions>
```

Any local "variables" are merely nullary functions.

Reply:
Hence no changes needed.

Reply:
I don't agree with this conclusion, but I'll open a separate PR/discussion after this is merged.